Add recipe for mimetypes

cniles opened the pull request:
### Brief summary of what the package does

mimetypes provides platform-dependent mechanisms for guessing MIME type by file extension

### Direct link to the package repository

https://github.com/cniles/emacs-mimetypes/

### Your association with the package

Maintainer & author

### Relevant communications with the upstream package maintainer

None needed

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

alphapapa commented:
Looks like you forgot to fill out the template.

cniles commented:
🤦 sorry; filled it out.

alphapapa commented:
Ok, here's a little feedback:

1.  Have you considered using the `file` utility on Unix systems?
2.  Simple functions that are only called in one place, like `mimetypes--line-has-extension`, should usually be inlined.

cniles commented:
@alphapapa, thanks, I appreciate you taking the time to do a thoughtful review.

1. No, the initial use case was for uploading buffers explicitly which may not have a visited file.  After some thought, I think there can be a reasonable argument for its use.  I've added two additional functions, one that handles just a file name and another that handles buffers.  In the case of a buffer, it will consider whether there is a visited file or not.  If so (and on a *nix system w/ file executable available) it will utilize `file` to try and determine a MIME type for the file.

2. I've inlined what seem to be simple functions that this feedback may apply to--noting the docs, however:

> ...you should not make a function inline, even if it’s small, unless its speed is really crucial, and you’ve timed the code to verify that using defun actually has performance problems. [0]

But I will defer to your experience.

3. I've refactored `mimetypes--find-in-buffer` to be a single function call to `re-search-forward` sans loop.

[0 https://www.gnu.org/software/emacs/manual/html_node/elisp/Inline-Functions.html](https://www.gnu.org/software/emacs/manual/html_node/elisp/Inline-Functions.html)

alphapapa commented:
> 1. In the case of a buffer, it will consider whether there is a visited file or not.  If so (and on a *nix system w/ file executable available) it will utilize `file` to try and determine a MIME type for the file.

According to the `file` man page, it can test STDIN, so you can use it to test non-file-backed buffers too:

```
     -f, --files-from namefile
             Read the names of the files to be examined from namefile (one per line) before the argument list.  Either namefile or at least one filename argument must be present; to test the standard input, use ‘-’ as a filename argument.
```

> I've inlined what seem to be simple functions that this feedback may apply to--noting the docs, however: "you should not make a function inline, even if it’s small, unless its speed is really crucial, and you’ve timed the code to verify that using defun actually has performance problems."

To be clear, by "inline" I mean to simply put the function's code where it is called, because since it's only called one time, and it's a one-line function, there's no benefit to putting it in a separate function.  I don't mean to use `define-inline` or `defsubst`.

> 3.  I've refactored mimetypes--find-in-buffer to be a single function call to re-search-forward sans loop.

Great.

alphapapa commented:
FYI I added some feedback as commit comments: https://github.com/cniles/emacs-mimetypes/commit/e02c36fbc57a17ce19309a987d8caf69b6fbd229

cniles commented:
@alphapapa believe I've made changes that take into account all your feedback.

alphapapa commented:
Cool.  FYI, I see 3 tests failing, but maybe it's just me?
```
Running 7 tests (2020-11-02 23:27:56-0600)
   passed  1/7  test-find-in-buffer
Test test-find-in-file backtrace:
  signal(ert-test-failed (((should (string= "text/plain" (mimetypes--f
  ert-fail(((should (string= "text/plain" (mimetypes--find-in-file "tx
  (if (unwind-protect (setq value-27 (apply fn-25 args-26)) (setq form
  (let (form-description-29) (if (unwind-protect (setq value-27 (apply
  (let ((value-27 (quote ert-form-evaluation-aborted-28))) (let (form-
  (let* ((fn-25 (function string=)) (args-26 (condition-case err (let 
  (closure (t) nil (let* ((fn-25 (function string=)) (args-26 (conditi
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name test-find-in-file :documentation nil 
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  command-line-1(("--title" "makem.sh: emacs-mimetypes (sandbox: /tmp/
  command-line()
  normal-top-level()
Test test-find-in-file condition:
    (ert-test-failed
     ((should
       (string= "text/plain"
                (mimetypes--find-in-file "txt" "one")))
      :form
      (string= "text/plain" nil)
      :value nil))
   FAILED  2/7  test-find-in-file
   passed  3/7  test-find-in-list
   passed  4/7  test-find-in-user-file
Test test-first-known-file backtrace:
  signal(ert-test-failed (((should (string= "one" (mimetypes--first-kn
  ert-fail(((should (string= "one" (mimetypes--first-known-file (quote
  (if (unwind-protect (setq value-32 (apply fn-30 args-31)) (setq form
  (let (form-description-34) (if (unwind-protect (setq value-32 (apply
  (let ((value-32 (quote ert-form-evaluation-aborted-33))) (let (form-
  (let* ((fn-30 (function string=)) (args-31 (condition-case err (let 
  (closure (t) nil (let* ((fn-30 (function string=)) (args-31 (conditi
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name test-first-known-file :documentation 
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  command-line-1(("--title" "makem.sh: emacs-mimetypes (sandbox: /tmp/
  command-line()
  normal-top-level()
Test test-first-known-file condition:
    (ert-test-failed
     ((should
       (string= "one"
                (mimetypes--first-known-file ...)))
      :form
      (string= "one" nil)
      :value nil))
   FAILED  5/7  test-first-known-file
Test test-guess-mime backtrace:
  signal(wrong-type-argument (stringp nil))
  apply(signal (wrong-type-argument (stringp nil)))
  (setq value-77 (apply fn-75 args-76))
  (unwind-protect (setq value-77 (apply fn-75 args-76)) (setq form-des
  (if (unwind-protect (setq value-77 (apply fn-75 args-76)) (setq form
  (let (form-description-79) (if (unwind-protect (setq value-77 (apply
  (let ((value-77 (quote ert-form-evaluation-aborted-78))) (let (form-
  (let* ((fn-75 (function string=)) (args-76 (condition-case err (let 
  (let ((mimetypes-known-files (quote ("one")))) (let* ((fn-75 (functi
  (closure (t) nil (let ((mimetypes-known-files (quote ("one")))) (let
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test
  ert-run-test(#s(ert-test :name test-guess-mime :documentation nil :b
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test 
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #
  ert-run-tests-batch(nil)
  ert-run-tests-batch-and-exit()
  command-line-1(("--title" "makem.sh: emacs-mimetypes (sandbox: /tmp/
  command-line()
  normal-top-level()
Test test-guess-mime condition:
    (wrong-type-argument stringp nil)
   FAILED  6/7  test-guess-mime
   passed  7/7  test-trim-extension

Ran 7 tests, 4 results as expected, 3 unexpected (2020-11-02 23:27:56-0600)

3 unexpected results:
   FAILED  test-find-in-file
   FAILED  test-first-known-file
   FAILED  test-guess-mime
```
I'm using https://github.com/alphapapa/makem.sh to run them, like:
`./makem.sh -vv --sandbox all`

cniles commented:
Looks like a working director issue; regardless, mimetypes-extension-to-mime shouldn't signal an error if a mime.types file can't be found.  Pushed a fix up for that.

cniles commented:
@alphapapa Unit tests are fixed to no longer be platform dependent and can run independent of working directory.

cniles commented:
@alphapapa anything other things holding this up?  @riscy additional thoughts?  If this is not reasonably innovative I'm not opposed to adding it in with the package it was originally intended for (axe).  I looked at `mailcap-extension-to-mime` and it seems to be lacking.  This package provides functionality similar to both PHP `mime_content_type`[1] and python `mimetypes`[2] with consideration for emacs-specific use cases.

[1] https://www.php.net/manual/en/function.mime-content-type.php
[2] https://docs.python.org/3/library/mimetypes.html

alphapapa commented:
> @alphapapa anything other things holding this up?

To be clear, I'm not a MELPA maintainer, just a fellow package author who helps by offering reviews sometimes.

> If this is not reasonably innovative I'm not opposed to adding it in with the package it was originally intended for (axe). I looked at mailcap-extension-to-mime and it seems to be lacking.

It seems generally useful and reasonably innovative to me.  If you're interested, I'd suggest offering it to Emacs upstream.  Of course, I'd expect them to have more feedback about how it should be written and integrate with existing Emacs code, so you may or may not feel like going through that.  But if it could be viewed as filling gaps in functionality, like in that function, it might be worth it.  The code doesn't seem that complex, and it might be more likely to get used if it were in Emacs rather than in a MELPA package that is likely to be overlooked among the other 5,000 packages.  Something to consider, anyway.  :)

riscy commented:
Thanks!  This looks great to me.  I think it's fine to merge it now, and if it were to be folded into Emacs in the future, we could remove it from the MELPA index.

tarsius commented:
@cniles Just a friendly reminder to revert https://github.com/cniles/axe/commit/a46e8e3199e13c76d98d0a6e3b55de71c4a80ecb and https://github.com/cniles/axe/commit/8f37e3e1dadcbb999f89ae0d75e6b121010ca1b1.
